### ISSUE TITLE
Job Declarator server fails with EmptyMempool

### ISSUE THREAD

**Issue by Sjors** [stratum-mining/stratum#744]
I run a custom signet, so sometimes the mempool is actually empty.

```
     Running `/home/sjors/dev/stratum/roles/target/debug/jd_server -c /home/sjors/.stratum/signet-jd-server.toml`
2024-02-06T11:36:50.014818Z  INFO jd_server: Jds INITIALIZING with config: "/home/sjors/.stratum/signet-jd-server.toml"
2024-02-06T11:36:50.014873Z  INFO jd_server::lib::job_declarator: JD INITIALIZED
2024-02-06T11:36:50.015496Z ERROR jd_server: EmptyMempool
2024-02-06T11:36:50.015507Z ERROR jd_server: Unable to connect to Template Provider (possible reasons: not fully synced, down)
2024-02-06T11:36:50.015528Z ERROR jd_server: SHUTDOWN from Upstream: Mempool error: `EmptyMempool`
```

But when I create a transaction so that there's something in the mempool, I still the same error.

**Comment by Sjors** [comment]
As a quick workaround I did:

```
                // if mempool_ordered.is_empty() {
                //     Err(JdsMempoolError::EmptyMempool)
                // } else {
                    Ok(mempool_ordered)
                // }
```

**Comment by lorbax** [comment]
It happened the same for me. It should be that much difficult to fix and I can do it in my PR right now

**Comment by Sjors** [comment]
It seems unrelated though, so maybe make a separate PR?

**Comment by Sjors** [comment]
Oh, I guess you mean #716 with "my PR". I thought you were talking about #737

**Comment by GitGab19** [comment]
> I run a custom signet, so sometimes the mempool is actually empty.
> 
> ```
>      Running `/home/sjors/dev/stratum/roles/target/debug/jd_server -c /home/sjors/.stratum/signet-jd-server.toml`
> 2024-02-06T11:36:50.014818Z  INFO jd_server: Jds INITIALIZING with config: "/home/sjors/.stratum/signet-jd-server.toml"
> 2024-02-06T11:36:50.014873Z  INFO jd_server::lib::job_declarator: JD INITIALIZED
> 2024-02-06T11:36:50.015496Z ERROR jd_server: EmptyMempool
> 2024-02-06T11:36:50.015507Z ERROR jd_server: Unable to connect to Template Provider (possible reasons: not fully synced, down)
> 2024-02-06T11:36:50.015528Z ERROR jd_server: SHUTDOWN from Upstream: Mempool error: `EmptyMempool`
> ```
> 
> But when I create a transaction so that there's something in the mempool, I still the same error.

I introduced that error management in my last PR (#733).
Before it, it was simply not managed and it caused a panic on JDS.
The point is that mempool should never be empty on mainnet, so if that error is generated it means that there are some issues with TP and connection to it should be closed. 
I also saw this strange behaviour during testing with asic, and I also thought about introducing a counter (which can be set in config) to be used before closing the connection.
But it depends on how we want to manage this specific case..

**Comment by Sjors** [comment]
You can always log a warning that the mempool is empty, but it's not a reason to crash. The mainnet mempool _can_ be empty. And it _is_ for a newly installed node, or if someone deleted `mempool.dat`.

**Comment by GitGab19** [comment]
> You can always log a warning that the mempool is empty, but it's not a reason to crash. The mainnet mempool _can_ be empty. And it _is_ for a newly installed node, or if someone deleted `mempool.dat`.

I can integrate this suggestion and open open a new PR. But it depends on how @lorbax is working on his PR #716. 
Let me know how you prefer to manage it.

**Comment by Sjors** [comment]
This bug is quite annoying, so I would prefer a simple quick fix until #716 is ready.

**Comment by lorbax** [comment]
I am rebasing [716](https://github.com/stratum-mining/stratum/pull/716) btw

**Comment by GitGab19** [comment]
> This bug is quite annoying, so I would prefer a simple quick fix until #716 is ready.

Just opened a PR to fix this issue (#747).
I already has a little testing and it seemed to work well, but a double check on this would be nice
